### DISCUSSION TITLE
단위 테스트, 통합 테스트, 슬라이스 테스트, 인수 테스트 등에 대해 설명해 주세요.

### DISCUSSION THREAD

**Discussion by 3Juhwan** [2024-woowacourse-study/level-interview#229] [Spring Test]
.

**Comment by jimi567** [comment]
### 단위 테스트 

+ 목적 : 가장 작은 단위의 코드(일반적으로 메소드) 가 올바르게 동작하는지를 확인하는 목적으로 작성된 테스트 입니다.
+ 특징 : 일반적으로 Junit을 사용하여 짜여진 테스트 입니다. 빠르게 실행되며, 코드 변경 시 가장 빠른 피드백을 제공합니다. 

###  통합 테스트

+ 목적 : 여러 개의 단위가 함께 동작함을 확인하기 위해 짜여진 테스트 입니다.
+ 특징 : 스프링에서는 리포지토리, 서비스 단을 포함한 실제 동작을 테스트 하기 위해 사용됩니다. 주로 스프링에서는 @SpringBootTest와 RestAssured를 통해 통합 테스트를 작성합니다.  
일반적으로 테스트에 소요되는 시간이 긴 편입니다. 

### 슬라이스 테스트

+ 목적 : 웹 애플리케이션에서 특정 레이어(컨트롤러, 리포지토리 등)에 대한 동작을 검증하기 위해 짜여지는 테스트 입니다.

+ 특징 : 레이어 별로 테스트 합니다. 주로 @WebMvcTest, @DataJpaTest 등을 통해 스프링에서 특정 레이어 동작에 필요한 빈만 등록하여 테스트할 수 있습니다. 따라서 통합 테스트에 비해 테스트 시간이 짧습니다. 

### 인수 테스트
+ 목적 : 사용자의 요구 사항이 충족 되었는지 확인하기 위한 테스트로, 주로 기획자가 만들어 둔 가상의 시나리오 대로 애플레케이션이 동작하는지를 테스트 합니다. 

+ 특징 : 통합 테스트와 유사하지만 시나리오 위주의 테스트로, 시나리오 테스트라고도 불리며, Selenium, Cucumber 등의 자동화된 테스트 도구를 사용하기도 합니다.

**Comment by GIVEN53** [comment]
### 단위 테스트
소프트웨어의 가장 작은 단위, 즉 개별 메서드나 함수의 기능을 검증하는 테스트입니다.
특정 기능이 올바르게 동작하는지 확인하기 위함이며 독립적이고 빠르게 실행됩니다,
또한 다른 시스템 컴포넌트나 외부 종속성에 의존하지 않고 종속성이 필요한 경우 모킹(Mock)을 통해 처리합니다.

### 통합 테스트
개별 모듈들이 결합되어 전체 시스템이 올바르게 동작하는지 검증하는 테스트입니다.
모듈 간의 상호작용이 올바르게 동작하는지 확인하기 위함이며 실제 데이터베이스, 네트워크 등의 외부 시스템과의 통합을 테스트하는 등 모듈 간의 인터페이스와 상호작용을 검증합니다.

### 슬라이스 테스트
Spring Boot에서 특정 레이어(ex. controller, service, repository)에 대한 테스트입니다.
목적: 애플리케이션의 특정 슬라이스(레이어)가 올바르게 동작하는지 확인하기 위함이며 스프링의 특정 컴포넌트만 로드하여 테스트합니다.
빠른 테스트를 위해 전체 애플리케이션 컨텍스트를 로드하지 않습니다.

### 인수테스트
시스템의 전체 기능을 사용자가 기대하는 방식으로 동작하는지 검증하는 테스트입니다.
시스템이 비즈니스 요구사항을 충족하는지 확인하기 위함이며 실제 사용자 시나리오를 기반으로 테스트합니다.
엔드 투 엔드(End-to-End, E2E) 테스트 형식이며 실제 애플리케이션 환경에서 수행됩니다.
